framework/src/org/checkerframework/framework/util/ContractsUtils.java

```diff
@@ -49,12 +50,154 @@ public static ContractsUtils getInstance(
     }
 
     /**
+     * Represents a pre- or postcondition that must be verified by
+     * {@code BaseTypeVisitor} or one of its subclasses. Automatically
+     * extracted from annotations with meta-annotations
+     * {@code @PreconditionAnnotation} or {@code @PostconditionAnnotation},
+     * such as {@code @RequiresNonNull} and {@code @EnsuresNonNull}.
+     * Can also be generated, such as in
+     * {@code LockVisitor.generatePreconditionsBasedOnGuards}.
+     */
+    public static class PreOrPostcondition {
+        /**
+         * The expression for which the condition must hold, such as
+         * {@code "foo"} in {@code @RequiresNonNull("foo")}.
+         */
+        public final String expression;
+        /**
+         * The name of the qualifier class that describes the condition that
+         * must hold.
+         */
+        public final String annotationString;
+
+        public PreOrPostcondition(String expression, String annotationString) {
+            this.expression = expression;
+            this.annotationString = annotationString;
+        }
+    }
+
+    /**
+     * An Iterable set of PreOrPostcondition entries.
+     * Implements the Iterable interface but not the Set interface.
+     */
+    public static class PreOrPostconditionSet implements Iterable<PreOrPostcondition> {
+        private Set<PreOrPostcondition> conditionSet;
+
+        private static final PreOrPostconditionSet EMPTY_SET =
+                new PreOrPostconditionSet(Collections.<PreOrPostcondition>emptySet());
+
+        private PreOrPostconditionSet(Set<PreOrPostcondition> initialSet) {
+            conditionSet = initialSet;
+        }
+
+        public PreOrPostconditionSet() {
+            conditionSet = new LinkedHashSet<PreOrPostcondition>(0);
+        }
+
+        public static PreOrPostconditionSet emptySet() {
+            return EMPTY_SET;
+        }
+
+        public void add(PreOrPostcondition p) {
+            conditionSet.add(p);
+        }
+
+        public void addAll(PreOrPostconditionSet set) {
+            conditionSet.addAll(set.conditionSet);
+        }
+
+        @Override
+        public Iterator<PreOrPostcondition> iterator() {
+            return conditionSet.iterator();
+        }
+
+        public boolean isEmpty() {
+            return conditionSet.isEmpty();
+        }
+    }
+
+    /**
+     * Represents a conditional postcondition that must be verified by
+     * {@code BaseTypeVisitor} or one of its subclasses. Automatically
+     * extracted from annotations with meta-annotation
+     * {@code @ConditionalPostconditionAnnotation}, such as
+     * {@code EnsuresNonNullIf}.
+     */
+    public static class ConditionalPostcondition {
```

Comment:
Should this be a subclass of `PreOrPostcondition`?
Maybe a note to keep the two classes in sync would be enough.